backend/src/main/java/com/google/sps/datastore/MatchDatastore.java

```diff
@@ -0,0 +1,70 @@
+package com.google.sps.datastore;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.sps.data.Match;
+
+public final class MatchDatastore {
+
+  /** Datastore */
+  private DatastoreService datastore;
+
+  // Datastore Key/Property constants
+  private static final String KIND_MATCH = "Match";
+  private static final String PROPERTY_FIRSTPARTICIPANTKEY = "firstParticipantKey";
+  private static final String PROPERTY_SECONDPARTICIPANTKEY = "secondParticipantKey";
+  private static final String PROPERTY_DURATION = "duration";
+  private static final String PROPERTY_TIMESTAMP = "timestamp";
+
+  /** Constructor */
+  public MatchDatastore(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  /** Add Match to datastore and return the match Key */
+  public Key addMatch(Match match) {
+    // Set properties of entity
+    Entity matchEntity = new Entity(KIND_MATCH);
+    matchEntity.setProperty(PROPERTY_FIRSTPARTICIPANTKEY, match.getFirstParticipantKey());
+    matchEntity.setProperty(PROPERTY_SECONDPARTICIPANTKEY, match.getSecondParticipantKey());
+    matchEntity.setProperty(PROPERTY_DURATION, match.getDuration());
+    matchEntity.setProperty(PROPERTY_TIMESTAMP, match.getTimestamp());
+
+    // Insert entity into datastore
+    datastore.put(matchEntity);
+
+    // Return matchKey
+    return matchEntity.getKey();
+  }
+
+  /** Return match based on match datastore key */
+  public Match getMatchFromKey(Key matchKey) {
+    // No match
+    if (matchKey == null) {
+      return null;
+    }
```

Comment:
Good opportunity here to start introducing nullability annotations to our java code.
https://mvnrepository.com/artifact/com.google.code.findbugs/jsr305/3.0.2

You can mark fields, parameters, and method return types as Nullable or NotNull using annotations. The compiler doesn't enforce these, but linters and other analysis tools can add warnings to code for example that passes null as an argument to a method where the parameter is marked as @NonNull. 

The annotations also help define the contract for your method. I would argue here that requesting something from a datastore using a null key is something the caller should guard against, not the datastore. I recommend adding a @NonNull annotation to the parameter and a @Nullable annotation to the method.

Reply:
Added @Nullable annotations to methods that return null. For the part you pointed out above, I realized I did check if it was null in the caller, so I removed the null check.